### PR TITLE
network: IPoIB support

### PR DESCRIPTION
**Pull request by yuwata**
This is the last part of #21255.

I have not tested the code, as I do not have Infiniband card... But at least, I hope this does not break other existing functionalities.

### CODE REVIEW

File: src/network/netdev/netdev.c [hunk 1]
```diff
@@ -430,21 +434,42 @@ int netdev_generate_hw_addr(
...
+                                                         "Failed to generate persistent MAC address, ignoring: %m");
+                                a = HW_ADDR_NULL;
+                                goto finalize;
+                        }
+
+                        break;
+                case ARPHRD_INFINIBAND:
+                        if (result == 0) {
+                                log_netdev_warning_errno(netdev, SYNTHETIC_ERRNO(EINVAL),
+                                                          "Failed to generate persistent MAC address: %m");
```
Comment (bluca) [inline src/network/netdev/netdev.c]: indentation looks off by one space?

File: src/network/netdev/ipoib.c [hunk 2]
```diff
@@ -56,6 +57,48 @@ static int netdev_ipoib_fill_message_create(NetDev *netdev, Link *link, sd_netli
...
+int ipoib_set_netlink_message(Link *link, sd_netlink_message *m) {
+        int r;
+
+        assert(link);
+        assert(link->network);
+        assert(m);
+
+        r = sd_netlink_message_set_flags(m, NLM_F_REQUEST | NLM_F_ACK);
+        if (r < 0)
+                return log_link_debug_errno(link, r, "Could not set netlink flags: %m");
```
Comment (bluca) [inline src/network/netdev/ipoib.c:69]: mention the flags in the message?

File: src/network/networkd-setlink.c [hunk 3]
```diff
@@ -600,6 +600,15 @@ static bool link_is_ready_to_call_set_link(Request *req) {
...
                 break;
         }
+        case SET_LINK_MTU: {
+                Request req_ipoib = {
+                        .link = link,
+                        .type = REQUEST_TYPE_SET_LINK,
+                        .set_link_operation_ptr = INT_TO_PTR(SET_LINK_IPOIB),
+                };
+
+                return !ordered_set_contains(link->manager->request_queue, &req_ipoib);
```
Comment (bluca) [inline src/network/networkd-setlink.c:611]: link->manager is not derefed in this function before this, add a check or assert in this block?

**Review by bluca — APPROVED** [review]
code looks fine, but I know really nothing about infiniband, is there someone with HW who might be able to test that this works?

File: src/network/netdev/ipoib.c (hunk 2 above)
Comment (yuwata) [reply src/network/netdev/ipoib.c:69]: This should not fail here, and if it fails, the error cause is not related to the flags, as the function accepts any non-zero flags.

**Comment by yuwata** [comment]
@bluca Thank you for your review. The above points are addressed, except for the log message for `sd_netlink_message_set_flags()`. Upgrading the green label.

> code looks fine, but I know really nothing about infiniband, is there someone with HW who might be able to test that this works?

I hope people who has HW will test and report if the code has any issues.

**Comment by bluca** [comment]
```
../src/network/netdev/ipoib.c: In function ‘ipoib_set_netlink_message’:
../src/network/netdev/ipoib.c:67:42: error: ‘req’ undeclared (first use in this function)
   67 |         r = sd_netlink_message_set_flags(req, NLM_F_REQUEST | NLM_F_ACK);
      |                                          ^~~
```

**Pull request merged**